File: backend/controllers/eventController.js
import pool from "../models/pool.js";
import { convertTagArray, getTrendingEvents } from "../utils/helpers.js";
import { v7 as uuidv7 } from "uuid";
import path from "path";
import logger from "../logger.js";

export const addEvents = async (req, res) => {
  const { events } = req.body;

  if (!Array.isArray(events)) {
    logger.error("Invalid input format");
    return res.status(400).json({ error: "Invalid input format" });
  }

  const client = await pool.connect();

  // Check that all events have a "header" field, as otherwise the database will not like it.
  const invalidEvents = events.filter(event => !event.header);
  if (invalidEvents.length > 0) {
    logger.error("Events missing required header field");
    return res.status(400).json({ error: "All events must have a header field" });
  }

  try {
    await client.query("BEGIN");

    const eventPromises = events.map(event => {
      const {
        header,
        link,
        source,
        admiralty_reliability,
        admiralty_accuracy,
        keywords,
        event_time,
        notes,
        hcoe_domains,
        location,
        location_lng,
        location_lat,
        author,
        groups
      } = event;
      const id = uuidv7();
      const keywordArray = convertTagArray(keywords);
      const groupsArray = groups ? (Array.isArray(groups) ? groups : [groups]) : [];

      return client.query(
        "INSERT INTO events (id, header, link, source, admiralty_reliability, admiralty_accuracy, keywords, event_time, notes, hcoe_domains, location, location_lng, location_lat, author, groups) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15)",
        [id, header, link, source, admiralty_reliability, admiralty_accuracy, keywordArray, event_time, notes, hcoe_domains, location, location_lng, location_lat, author, groupsArray]
      );
    });

    await Promise.all(eventPromises);

    await client.query("COMMIT");
    logger.info("Events added successfully");
    res.status(201).json({ message: "Events added successfully" });
  } catch (error) {
    await client.query("ROLLBACK");
    logger.error("ROLLBACK: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const fetchEvents = async (req, res) => {
  const { search } = req.query;

  const client = await pool.connect();
  try {
    let query = "SELECT * FROM events";
    const values = [];

    if (search) {
      const searchWords = search.split(" ").map(word => word.trim()).filter(Boolean);

      if (searchWords.length > 0) {
        const conditions = searchWords.map((_, index) => `
                    (
                        EXISTS (SELECT 1 FROM unnest(keywords) AS keyword WHERE keyword ILIKE $${index + 1}) OR
                        header ILIKE $${index + 1} OR
                        notes ILIKE $${index + 1} OR
                        EXISTS (SELECT 1 FROM unnest(hcoe_domains) AS domain WHERE domain ILIKE $${index + 1}) OR
                        link ILIKE $${index + 1} OR
                        source ILIKE $${index + 1}
                    )
                `);

        query += ` WHERE ${conditions.join(" AND ")}`;
        values.push(...searchWords.map(word => `%${word}%`));
      }
    }

    query += " ORDER BY creation_time DESC";

    const result = await client.query(query, values);

    res.json(result.rows);
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};


export const fetchEventById = async (req, res) => {
  const client = await pool.connect();
  const eventId = req.params.id;

  try {
    const eventResult = await client.query("SELECT * FROM events WHERE id = $1", [eventId]);
    if (eventResult.rows.length === 0) {
      return res.status(404).json({ error: "Event not found" });
    }

    const event = eventResult.rows[0];

    const imagesResult = await client.query("SELECT url FROM event_media WHERE event_id = $1", [eventId]);
    const images = imagesResult.rows.map(row => row.url);

    res.json({ ...event, images });
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const fetchTrendingEventsDay = async (req, res) => {
  const client = await pool.connect();
  try {
    const result = await client.query(
      `SELECT *
       FROM events
       WHERE creation_time >= NOW() - INTERVAL '24 hours'`
    );

    const {
      trendingEvents,
      trendingTag,
      percentageOfTotalTags,
      numberOfTags
    } = getTrendingEvents(result.rows);
    res.json({
      trending_tag: trendingTag,
      precentage_of_tags: percentageOfTotalTags,
      number_of_tags: numberOfTags,
      events: trendingEvents
    });
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const fetchTrendingEventsWeek = async (req, res) => {
  const client = await pool.connect();
  try {
    const result = await client.query(
      `SELECT *
       FROM events
       WHERE creation_time >= NOW() - INTERVAL '1 week'`
    );

    const {
      trendingEvents,
      trendingTag,
      percentageOfTotalTags,
      numberOfTags
    } = getTrendingEvents(result.rows);
    res.json({
      trending_tag: trendingTag,
      precentage_of_tags: percentageOfTotalTags,
      number_of_tags: numberOfTags,
      events: trendingEvents
    });
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const fetchKeywords = async (req, res) => {
  const client = await pool.connect();
  try {
    const result = await client.query(`
      SELECT UNNEST(keywords) AS keyword, COUNT(*) AS count
      FROM events
      GROUP BY keyword
      ORDER BY count DESC
    `);

    const keywordsWithCount = {};
    for (const row of result.rows) {
      keywordsWithCount[row.keyword] = parseInt(row.count, 10);
    }

    res.json(keywordsWithCount);
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const searchEventsByLocation = async (req, res) => {
  const { longitude, latitude, radius } = req.query;

  if (!longitude || !latitude || !radius) {
    return res.status(400).json({ error: "Longitude, latitude, and radius are required" });
  }

  const client = await pool.connect();
  try {
    const query = `
      SELECT *
      FROM events
      WHERE ST_DWithin(
              ST_SetSRID(ST_MakePoint($1, $2), 4326) ::geography,
              ST_SetSRID(ST_MakePoint(location_lng, location_lat), 4326) ::geography,
              $3 * 1000
            )
      ORDER BY creation_time DESC
    `;
    const values = [longitude, latitude, radius];

    const result = await client.query(query, values);
    res.json(result.rows);
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const uploadImages = async (req, res) => {
  const { eventId } = req.body;
  const files = req.files;

  if (!eventId || !files || files.length === 0) {
    return res.status(400).json({ error: "Event ID and image files are required" });
  }

  const client = await pool.connect();
  try {
    await client.query("BEGIN");

    const uploadPromises = files.map(file => {
      const id = uuidv7();
      const url = path.join("uploads", file.filename);
      return client.query(
        "INSERT INTO event_media (id, event_id, url) VALUES ($1, $2, $3)",
        [id, eventId, url]
      );
    });

    await Promise.all(uploadPromises);

    await client.query("COMMIT");
    res.status(201).json({
      message: "Images uploaded successfully",
      files: files.map(file => file.filename)
    });
  } catch (error) {
    await client.query("ROLLBACK");
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

export const fetchMetrics = async (req, res) => {
  const client = await pool.connect();
  try {
    // Number of events
    const eventsCountResult = await client.query("SELECT COUNT(*) FROM events");
    const eventsCount = parseInt(eventsCountResult.rows[0].count, 10);

    // Number of unique keywords
    const uniqueKeywordsResult = await client.query("SELECT COUNT(DISTINCT keyword) FROM (SELECT UNNEST(keywords) AS keyword FROM events) AS unique_keywords");
    const uniqueKeywordsCount = parseInt(uniqueKeywordsResult.rows[0].count, 10);

    // Total number of keywords
    const totalKeywordsResult = await client.query("SELECT COUNT(keyword) FROM (SELECT UNNEST(keywords) AS keyword FROM events) AS total_keywords");
    const totalKeywordsCount = parseInt(totalKeywordsResult.rows[0].count, 10);

    // Count of every keyword
    const keywordsCountResult = await client.query(`
      SELECT keyword, COUNT(*) AS count
      FROM (SELECT UNNEST(keywords) AS keyword FROM events) AS keyword_counts
      GROUP BY keyword
      ORDER BY count DESC
    `);
    // eslint-disable-next-line unicorn/no-array-reduce
    const keywordsCount = keywordsCountResult.rows.reduce((acc, row) => {
      acc[row.keyword] = parseInt(row.count, 10);
      return acc;
    }, {});

    res.json({
      eventsCount,
      uniqueKeywordsCount,
      totalKeywordsCount,
      keywordsCount
    });
  } catch (error) {
    logger.error("Error: " + error.message);
    res.status(500).json({ error: error.message });
  } finally {
    client.release();
  }
};

// Group management functions
export const createGroup = async (req, res) => {
    const { eventIds, groupName } = req.body;

    if (!eventIds || !Array.isArray(eventIds) || eventIds.length === 0 || !groupName) {
        return res.status(400).json({ error: 'Event IDs array and group name are required' });
    }

    const client = await pool.connect();
    try {
        await client.query('BEGIN');

        // Add the group name to the groups array for all specified events
        const updateQuery = `
            UPDATE events 
            SET groups = CASE 
                WHEN groups IS NULL THEN ARRAY[$1]
                ELSE array_append(groups, $1)
            END
            WHERE id = ANY($2) AND NOT ($1 = ANY(groups))
        `;
        await client.query(updateQuery, [groupName, eventIds]);

        await client.query('COMMIT');
        logger.info(`Group "${groupName}" added to ${eventIds.length} events`);
        res.status(201).json({ message: `Group "${groupName}" added successfully`, groupName, eventCount: eventIds.length });
    } catch (error) {
        await client.query('ROLLBACK');
        logger.error('Error creating group: ' + error.message);
        res.status(500).json({ error: error.message });
    } finally {
        client.release();
    }
};

export const updateEventGroup = async (req, res) => {
    const { eventId } = req.params;
    const { groupName } = req.body;

    if (!groupName) {
        return res.status(400).json({ error: 'Group name is required' });
    }

    const client = await pool.connect();
    try {
        const result = await client.query(`
            UPDATE events 
            SET groups = CASE 
                WHEN groups IS NULL THEN ARRAY[$1]
                ELSE array_append(groups, $1)
            END
            WHERE id = $2 AND NOT ($1 = ANY(groups))
            RETURNING *
        `, [groupName, eventId]);

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Event not found or already in group' });
        }

        logger.info(`Event ${eventId} added to group "${groupName}"`);
        res.json({ message: 'Event group updated successfully', event: result.rows[0] });
    } catch (error) {
        logger.error('Error updating event group: ' + error.message);
        res.status(500).json({ error: error.message });
    } finally {
        client.release();
    }
};

export const fetchGroups = async (req, res) => {
    const client = await pool.connect();
    try {
        const result = await client.query(`
            SELECT unnest(groups) as group_name, COUNT(*) as event_count
            FROM events 
            WHERE groups IS NOT NULL AND array_length(groups, 1) > 0
            GROUP BY group_name
            ORDER BY group_name
        `);

        res.json(result.rows);
    } catch (error) {
        logger.error('Error fetching groups: ' + error.message);
        res.status(500).json({ error: error.message });
    } finally {
        client.release();
    }
};

export const fetchEventsByGroup = async (req, res) => {
    const { groupName } = req.params;

    const client = await pool.connect();
    try {
        const result = await client.query('SELECT * FROM events WHERE $1 = ANY(groups) ORDER BY creation_time DESC', [groupName]);

        if (result.rows.length === 0) {
            return res.status(404).json({ message: `No events found in group "${groupName}"` });
        }

        res.json(result.rows);
    } catch (error) {
        logger.error('Error fetching events by group: ' + error.message);
        res.status(500).json({ error: error.message });
    } finally {
        client.release();
    }
};

export const removeFromGroup = async (req, res) => {
    const { eventId } = req.params;
    const { groupName } = req.body;

    if (!groupName) {
        return res.status(400).json({ error: 'Group name is required' });
    }

    const client = await pool.connect();
    try {
        const result = await client.query(`
            UPDATE events 
            SET groups = array_remove(groups, $1)
            WHERE id = $2
            RETURNING *
        `, [groupName, eventId]);

        if (result.rows.length === 0) {
            return res.status(404).json({ error: 'Event not found' });
        }

        logger.info(`Event ${eventId} removed from group "${groupName}"`);
        res.json({ message: 'Event removed from group successfully', event: result.rows[0] });
    } catch (error) {
        logger.error('Error removing event from group: ' + error.message);
        res.status(500).json({ error: error.message });
    } finally {
        client.release();
    }
};

File: backend/rust/src/controllers/event_controller.rs
use crate::db::DbPool;
use crate::models::event::*;
use crate::schema::events;
use diesel::prelude::*;
use rocket::State;
use serde_json::{json, Value};
use std::collections::HashMap;
use uuid::Uuid;

pub async fn fetch_events(db: &State<DbPool>) -> Value {
    use crate::schema::events::dsl::*;

    let mut conn = db.get().expect("Failed to get DB connection");

    match events
        .order_by(creation_time.desc())
        .select(Event::as_select())
        .get_results(&mut conn)
    {
        Ok(results) => {
            if results.is_empty() {
                json!([])
            } else {
                json!(results)
            }
        }
        Err(e) => {
            eprintln!("Database error: {}", e);
            json!({
                "message": "Failed to fetch events from database"
            })
        }
    }
}

pub async fn add_events(event_data: Value, db: &State<DbPool>) -> Value {
    // Validate that event_data is an array
    if !event_data.is_array() {
        return json!({
            "error": "Invalid input format: expected an array of events"
        });
    }

    let events = event_data.as_array().unwrap();

    let mut conn = db.get().expect("Failed to get DB connection");

    // Start a transaction
    let result = conn.transaction(|conn| {
        for event in events {
            // Extract values from the event JSON
            let header = event
                .get("header")
                .and_then(|v| v.as_str())
                .unwrap_or_default()
                .to_string();
            let link = event
                .get("link")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let source = event
                .get("source")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let admiralty_reliability = event
                .get("admiralty_reliability")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let admiralty_accuracy = event
                .get("admiralty_accuracy")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let notes = event
                .get("notes")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let location = event
                .get("location")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let location_lng = event
                .get("location_lng")
                .and_then(|v| v.as_f64())
                .map(|v| v as f32);
            let location_lat = event
                .get("location_lat")
                .and_then(|v| v.as_f64())
                .map(|v| v as f32);
            let author = event
                .get("author")
                .and_then(|v| v.as_str())
                .map(|s| s.to_string());
            let event_time_str = event.get("event_time").and_then(|v| v.as_str());

            // Convert keywords to Vec<String>
            let keywords = match event.get("keywords") {
                Some(k) => {
                    let mut keyword_array: Vec<String> = Vec::new();
                    if let Some(k_str) = k.as_str() {
                        keyword_array = k_str
                            .split(',')
                            .map(|s| s.trim().to_string())
                            .filter(|s| !s.is_empty())
                            .collect();
                    } else if let Some(k_arr) = k.as_array() {
                        keyword_array = k_arr
                            .iter()
                            .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
                            .filter(|s| !s.is_empty())
                            .collect();
                    }
                    Some(keyword_array)
                }
                None => None,
            };

            // Convert HCOE domains to Vec<String>
            let hcoe_domains = match event.get("hcoe_domains") {
                Some(d) => {
                    if let Some(d_arr) = d.as_array() {
                        Some(
                            d_arr
                                .iter()
                                .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
                                .filter(|s| !s.is_empty())
                                .collect::<Vec<String>>(),
                        )
                    } else {
                        None
                    }
                }
                None => None,
            };

            // Parse event_time if provided and convert to string
            let event_time = match event_time_str {
                Some(time_str) if !time_str.trim().is_empty() => {
                    match chrono::NaiveDateTime::parse_from_str(time_str, "%Y-%m-%dT%H:%M") {
                        Ok(dt) => Some(dt.format("%Y-%m-%dT%H:%M").to_string()), // Convert to string
                        Err(_) => None,
                    }
                }
                _ => None,
            };

            // Generate UUID for the new event
            let id = Uuid::now_v7();

            // Insert the event into the database
            diesel::insert_into(events::table)
                .values((
                    events::id.eq(id),
                    events::header.eq(header),
                    events::link.eq(link),
                    events::source.eq(source),
                    events::admiralty_reliability.eq(admiralty_reliability),
                    events::admiralty_accuracy.eq(admiralty_accuracy),
                    events::keywords.eq(keywords),
                    events::event_time.eq(event_time),
                    events::notes.eq(notes),
                    events::hcoe_domains.eq(hcoe_domains),
                    events::location.eq(location),
                    events::location_lng.eq(location_lng),
                    events::location_lat.eq(location_lat),
                    events::author.eq(author),
                    events::creation_time.eq(chrono::Utc::now()),
                ))
                .execute(conn)?;
        }

        diesel::result::QueryResult::Ok(())
    });

    // Return appropriate response based on transaction result
    match result {
        Ok(_) => json!({
            "message": "Events added successfully"
        }),
        Err(e) => json!({
            "error": format!("{}", e)
        }),
    }
}

pub async fn fetch_trending_events_day(db: &State<DbPool>) -> Value {
    // Implement trending events (day) logic here
    json!({
        "status": "success",
        "events": []
    })
}

pub async fn fetch_trending_events_week(db: &State<DbPool>) -> Value {
    // Implement trending events (week) logic here
    json!({
        "status": "success",
        "events": []
    })
}

pub async fn fetch_event_by_id(id: String, db: &State<DbPool>) -> Value {
    // Implement fetch by ID logic here
    json!({
        "status": "success",
        "event": null
    })
}

pub async fn fetch_keywords(db: &State<DbPool>) -> Value {
    let mut conn = match db.get() {
        Ok(conn) => conn,
        Err(_) => {
            return json!({
                "error": "Failed to get database connection"
            })
        }
    };

    // Use raw SQL to unnest keywords array and count occurrences
    let results = diesel::sql_query(
        r#"
        SELECT UNNEST(keywords) AS keyword, COUNT(*) AS count
        FROM events
        WHERE keywords IS NOT NULL
        GROUP BY keyword
        ORDER BY count DESC
        "#,
    )
        .load::<KeywordCount>(&mut conn);

    match results {
        Ok(keyword_counts) => {
            // Convert to HashMap<String, i32> format like the JavaScript version
            let mut keywords_with_count = HashMap::new();
            for kc in keyword_counts {
                keywords_with_count.insert(kc.keyword, kc.count);
            }
            json!(keywords_with_count)
        }
        Err(e) => {
            eprintln!("Database error fetching keywords: {}", e);
            json!({
                "error": "Failed to fetch keywords from database"
            })
        }
    }
}

pub async fn search_events_by_location(query: Value, db: &State<DbPool>) -> Value {
    // Extract parameters
    let lat = match query.get("lat").and_then(|v| v.as_f64()) {
        Some(val) => val,
        None => {
            return json!({
                "status": "error",
                "message": "Missing or invalid latitude parameter"
            })
        }
    };

    let lng = match query.get("lng").and_then(|v| v.as_f64()) {
        Some(val) => val,
        None => {
            return json!({
                "status": "error",
                "message": "Missing or invalid longitude parameter"
            })
        }
    };

    let radius_km = query.get("radius").and_then(|v| v.as_f64()).unwrap_or(10.0);

    let mut conn = match db.get() {
        Ok(conn) => conn,
        Err(_) => {
            return json!({
                "status": "error",
                "message": "Failed to get database connection"
            })
        }
    };

    // Using PostGIS ST_DWithin for more efficient spatial query
    let results = diesel::sql_query(
        r#"
        SELECT * FROM events
        WHERE
            (location_lat IS NOT NULL) AND
            (location_lng IS NOT NULL) AND
            ST_DWithin(
                ST_SetSRID(ST_MakePoint(location_lng, location_lat), 4326),
                ST_SetSRID(ST_MakePoint($2, $1), 4326),
                $3 * 1000  -- Convert km to meters
            )
        ORDER BY
            ST_Distance(
                ST_SetSRID(ST_MakePoint(location_lng, location_lat), 4326),
                ST_SetSRID(ST_MakePoint($2, $1), 4326)
            ) ASC,
            creation_time DESC
    "#,
    )
    .bind::<diesel::sql_types::Double, _>(lat)
    .bind::<diesel::sql_types::Double, _>(lng)
    .bind::<diesel::sql_types::Double, _>(radius_km)
    .load::<Event>(&mut conn);

    match results {
        Ok(found_events) => {
            if found_events.is_empty() {
                json!([])
            } else {
                json!(found_events)
            }
        }
        Err(e) => {
            eprintln!("Database error searching by location: {}", e);
            json!({
                "status": "error",
                "message": "Failed to search events by location"
            })
        }
    }
}

pub async fn fetch_metrics(db: &State<DbPool>) -> Value {
    // Implement metrics fetch logic
    json!({
        "status": "success",
        "metrics": {}
    })
}
